BinaryProtocolExtensions.md

```diff
@@ -0,0 +1,133 @@
+    4 bytes   | PAR1
+
+After the migration period, the end of the Parquet file may look like this:
+
+    K bytes   | Flatbuffers representation (v1) of FileMetaData
+    4 bytes   | little-endian crc32(flatbuffer)
+    4 bytes   | little-endian size(flatbuffer)
+    4 bytes   | little-endian crc32(size(flatbuffer))
+    4 bytes   | PAR3
+
+In this example, we see several design decisions for the extension at play:
+
+* There is a new some-other-UUID for the accepted change to the standard and now the Thrift `FileMetaData` cannot be extended itself.  
+* The length of the footer and the crc32 of the length itself, guarantees that new readers will not overshoot reading bytes in case of corrupt bits in these critical 8 bytes of the file.  
+* The crc32 of the flatbuffer representation enhances Parquet to have crc32 for metadata as well which is arguably more important than crc32 for data.  
+* The new encoding itself, which MUST contain some way to be extended in the future (much like Thrift does with this specification).
+
+### Encoding
+
+The community experiments with a new encoding extension. At the same time they want to keep the newly encoded Parquet files open for everyone to read. So they add a new encoding via an extension to the `ColumnMetaData` struct. The extension stores offsets in the Parquet file where the new and duplicate encoded data for this column lives. The new writer carefully places all the new encodings at the start of the row group and all the old encodings at the end of the row group. This layout minimizes disruption for readers unaware of the new encodings.
+
+In its private form Parquet files look like so:
+
+    4 bytes   | PAR1
+              |             | Column b (new encoding)
+              |             | Column c (new encoding)
+    R bytes   |  Row Group  | Column a
+              |     0       | Column d
+              |             | Column b (old encoding)
+              |             | Column c (old encoding)
+              |             | FileMetaData
+              |             | ColumnMetaData: a
+              |             | ColumnMetaData: b
+    F bytes   |             | <extension-blob with offsets to new encoding>
+              |             | ColumnMetaData: c
+              |             | <extension-blob with offsets to new encoding>
+              |             | ColumnMetaData: d
+    4 bytes   | PAR1
+
+The custom reader is compiled with thrift IDL with a binary for field with id 32767. This is done to become extension aware and inspect the extension bytes looking for the UUID disambiguator. If that’s found it decodes the offsets from the rest of the bytes and reads the region of the file containing the new encoding.
+
+If/when the encoding is ratified, it is added to the official specification as an additional type in `Encodings` at which point the extension is no longer necessary, nor the duplicated data in the row group.
+
+## Appending extensions to thrift
```

Comment:
Have you considered directly adding a `32767: optional binary reserved_extension` field to `FileMetaData` and `ColumnMetaData` to make it easier for implementations to append data?

Reply:
Yes I have considered it. The advantage of not adding it is that readers will not materialize the extension string. This means all but the readers that care about the extension will not use additional memory or incur extra allocation because of it.